extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/CustomQueryMethodsAdder.java

```diff
@@ -0,0 +1,271 @@
+            String[] methodParameterTypesStr = new String[methodParameterTypes.size()];
+            List<Integer> queryParameterIndexes = new ArrayList<>(methodParameterTypes.size());
+            Integer pageableParameterIndex = null;
+            Integer sortParameterIndex = null;
+            for (int i = 0; i < methodParameterTypes.size(); i++) {
+                DotName parameterType = methodParameterTypes.get(i).name();
+                methodParameterTypesStr[i] = parameterType.toString();
+                if (DotNames.SPRING_DATA_PAGEABLE.equals(parameterType)
+                        || DotNames.SPRING_DATA_PAGE_REQUEST.equals(parameterType)) {
+                    if (pageableParameterIndex != null) {
+                        throw new IllegalArgumentException("Method " + method.name() + " of Repository " + repositoryClassInfo
+                                + "has invalid parameters - only a single parameter of type" + DotNames.SPRING_DATA_PAGEABLE
+                                + " can be specified");
+                    }
+                    pageableParameterIndex = i;
+                } else if (DotNames.SPRING_DATA_SORT.equals(parameterType)) {
+                    if (sortParameterIndex != null) {
+                        throw new IllegalArgumentException("Method " + method.name() + " of Repository " + repositoryClassInfo
+                                + "has invalid parameters - only a single parameter of type" + DotNames.SPRING_DATA_SORT
+                                + " can be specified");
+                    }
+                    sortParameterIndex = i;
+                } else if (!useNamedParams) {
+                    queryParameterIndexes.add(i);
+                }
+            }
+
+            // go through the method annotations, find the @Param annotation on parameters
+            // and map the name to the method param index
+            Map<String, Integer> namedParameterToIndex = new HashMap<>();
+            List<AnnotationInstance> annotations = method.annotations();
+            for (AnnotationInstance annotation : annotations) {
+                if ((annotation.target().kind() != AnnotationTarget.Kind.METHOD_PARAMETER)
+                        || (!DotNames.SPRING_DATA_PARAM.equals(annotation.name()))) {
+                    continue;
+                }
+                namedParameterToIndex.put(annotation.value().asString(),
+                        (int) annotation.target().asMethodParameter().position());
+            }
+
+            boolean isModifying = (method.annotation(DotNames.SPRING_DATA_MODIFYING) != null);
+            if (isModifying && (sortParameterIndex != null || pageableParameterIndex != null)) {
+                throw new IllegalArgumentException(
+                        method.name() + " of Repository " + repositoryClassInfo
+                                + " is meant to be a delete query and therefore doesn't " +
```

Comment:
Hmmm, update and insert queries are modifying too, aren't they?

Reply:
Absolutely correct